File: src/python/ipc/pubsub.py
import asyncio
import functools
import pathlib
import pickle
import time
from multiprocessing.shared_memory import SharedMemory
from typing import Callable, Generic, Optional, TypeVar

import filelock
import log
from typing_helpers import req

from ipc import core

_LOCK_DIR = "/home/brianlerner/beachbot/var/locks/"
_LOCK_SUFFIX = "-lock"

# Minumum interval in seconds between when checking for a new message.
_POLL_INTERVAL = 0.001

BaseMessageT = TypeVar("BaseMessageT", bound="core.BaseMessage")


class Publisher(Generic[BaseMessageT]):
    """A generic publisher for sending messages from a specific channel."""

    def __init__(self, node_id: core.NodeID, channel: core.ChannelSpec):
        self._node_id = node_id
        self._channel = channel
        self._shm = _init_shm(channel)
        # Should always be non-None because we just init the SHM file.
        self._shm_id = req(_get_shm_id(self._shm))
        self._lock = _gen_lock(self._shm.name)

    def publish(self, msg: BaseMessageT) -> None:
        """Sends a message to the channel.

        NOTE: Messages are not queued and only the latest message can be
        read by the subscriber.
        """
        with self._lock:
            if (cur_shm_id := _get_shm_id(self._shm)) is None:
                log.error(
                    f"{self._node_id} is attempting to publish to SHM {self._shm.name} which is closed."
                )
                return

            _raise_if_shm_id_changed(self._shm_id, cur_shm_id)
            self._send_message(msg)

    def close(self) -> None:
        """Stops the publisher, closing any open data links permanently."""
        with self._lock:
            _close(self._shm)

    def _send_message(self, msg: BaseMessageT) -> None:
        msg.origin = self._node_id
        msg.creation = time.perf_counter()
        serialized_msg = pickle.dumps(msg)
        self._raise_if_message_too_large(serialized_msg)
        self._write_to_shm(serialized_msg)

    def _write_to_shm(self, data: bytes) -> None:
        """Writes to SHM with a buffer if the message is not the exact size of the
        registry.
        """
        buffer = b"\x00" * (self._channel.msg_size - len(data))
        data_w_buffer = data + buffer
        self._shm.buf[: self._channel.msg_size] = data_w_buffer

    def _raise_if_message_too_large(self, msg: bytes) -> None:
        if len(msg) > self._channel.msg_size:
            raise ValueError(
                "Message is larger than the channel size. "
                f"Message size: {len(msg)}, "
                f"Channel size: {self._channel.msg_size}"
            )


class Subscriber(Generic[BaseMessageT]):
    """A generic subscriber for receiving messages sent to a specific channel. Once
    received, the message is passed to a callback function for processing.
    """

    def __init__(
        self,
        node_id: core.NodeID,
        channel: core.ChannelSpec[BaseMessageT],
        callback: Optional[Callable],
    ):
        self._node_id = node_id
        self._shm = _init_shm(channel)
        self._callback = callback
        # Should always be non-None because we just init the SHM file.
        self._shm_id = req(_get_shm_id(self._shm))
        self._lock = _gen_lock(self._shm.name)

        # Prime last message so we only return new messages
        self._last_msg: Optional[BaseMessageT] = self._get_msg()
        self._running = True

    async def wait_for_message(self) -> Optional[BaseMessageT]:
        """Waits for a message to be returned. Returns None in cases where the
        subscriber stops running or no further messages are expected due to the channel
        closing.
        """
        while self._running:
            if (cur_shm_id := _get_shm_id(self._shm)) is None:
                # If the current shm switches to None then we know that some other
                # process has closed and unlinked the shm file so we choose to just exit
                # the listen function vs raise.
                self._running = False
                log.info("Shutting down listen as the SHM file is unlinked.")
                return None

            _raise_if_shm_id_changed(self._shm_id, cur_shm_id)
            if (msg := self._get_msg()) is None or msg == self._last_msg:
                # Small sleep to let other concurrent processes run and not block.
                await asyncio.sleep(_POLL_INTERVAL)
            else:
                self._last_msg = msg
                return msg

        return None

    async def listen(self) -> None:
        """Listens indefinitely for the latest unique message and processes it via the
        configured callback.
        """
        if self._callback is None:
            raise ValueError("Callback must be specified to listen for messages.")

        while msg := await self.wait_for_message():
            if _is_coroutine(self._callback):
                await self._callback(msg)
            else:
                self._callback(msg)
                # Gives up the thread temporarily for regular callbacks to let other
                # concurrent processes run.
                await asyncio.sleep(0.0)

    def close(self) -> None:
        """Stops the subscriber for running, permanently closing any open links to
        the channel.
        """
        self._running = False
        with self._lock:
            _close(self._shm)

    def _get_msg(self) -> Optional[BaseMessageT]:
        try:
            # Copy the data to help avoid race conditions. Force an unpickling
            # error if the buffer is none.
            with self._lock:
                data = bytes(self._shm.buf or b"\x00")

            msg = pickle.loads(data)

        # Pickling errors can happen if the shm has no data. Set message to None.
        except pickle.UnpicklingError:
            msg = None

        return msg


def _close(shm: SharedMemory) -> None:
    """Closes the SHM by first calling close then unlink to permanently
    release the memory.
    """
    try:
        shm.close()
    except BufferError:
        pass

    try:
        shm.unlink()
    except FileNotFoundError:
        pass


def _get_shm_id(shm: SharedMemory) -> Optional[int]:
    """The SHM id as represented by a unique identifier of the file path."""
    try:
        return _shm_path(shm).stat().st_ino
    except FileNotFoundError:
        return None


@functools.lru_cache
def _shm_path(shm: SharedMemory) -> pathlib.Path:
    return pathlib.Path("/dev/shm") / shm.name


def _init_shm(channel: core.ChannelSpec) -> SharedMemory:
    name = channel.name()
    try:
        shm = SharedMemory(name, create=True, size=channel.msg_size)
    except FileExistsError:
        shm = SharedMemory(name)

    return shm


def _gen_lock(shm_name: str) -> filelock._unix.UnixFileLock:
    """Generates a thread agnostic filelock for the given shm name used to avoid race
    conditions.
    """
    lock_path = _LOCK_DIR + shm_name + _LOCK_SUFFIX
    return filelock._unix.UnixFileLock(lock_path, thread_local=False)


@functools.lru_cache
def _is_coroutine(callback: Callable) -> bool:
    return asyncio.iscoroutinefunction(functools._unwrap_partial(callback))  # type: ignore[attr-defined]


def _raise_if_shm_id_changed(shm_id_1: int, shm_id_2: int) -> None:
    if shm_id_1 != shm_id_2:
        raise ValueError(f"SHM ID has changed, file closed. {shm_id_1=}, {shm_id_2=}")

File: src/cpp/ipc/pubsub.cpp
#include <iostream>
#include <string>
#include <sys/mman.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <cstring>
#include <stdexcept>
#include <chrono>
#include <thread>
#include <functional>
#include <optional>
#include <fstream>
#include <atomic>
#include <sys/file.h>
#include <cerrno>
#include "core.h"
#include "pubsub.h"

using namespace std;

// Constants
const string LOCK_DIR = "/tmp/";  // Adjust for macOS
const string LOCK_SUFFIX = "-lock";
const chrono::duration<double> POLL_INTERVAL(0.001);

// Forward decl helpers
static int get_shm_id(int shm_fd);
static void close_shm(int shm_fd, const std::string& shm_name);
static int init_shm(ChannelSpec& channel);
static int gen_lock(const std::string& shm_name);

// Publisher implementation
Publisher::Publisher(const NodeID& node_id, const ChannelSpec& channel)
    : _node_id(node_id), _channel(channel) {
    _shm_fd = init_shm(_channel);
    _shm_ptr = mmap(nullptr, _channel.msg_size, PROT_READ | PROT_WRITE, MAP_SHARED, _shm_fd, 0);
    if (_shm_ptr == MAP_FAILED) throw runtime_error("mmap failed");
    _shm_id = get_shm_id(_shm_fd);
    _lock_fd = gen_lock(_channel.name());
}

Publisher::~Publisher() {
    if (_shm_ptr) munmap(_shm_ptr, _channel.msg_size);
    if (_shm_fd != -1) ::close(_shm_fd);
}

void Publisher::publish(BaseMessage& msg) {
    flock(_lock_fd, LOCK_EX);
    int cur_shm_id = get_shm_id(_shm_fd);
    if (cur_shm_id == -1) {
        flock(_lock_fd, LOCK_UN);
        throw runtime_error("SHM is closed");
    }
    if (cur_shm_id != _shm_id) {
        flock(_lock_fd, LOCK_UN);
        throw runtime_error("SHM ID changed");
    }
    // Serialize
    msg.origin = _node_id;
    msg.creation = chrono::duration_cast<chrono::duration<double>>(chrono::high_resolution_clock::now().time_since_epoch()).count();
    string serialized;
    double creation_val = msg.creation.value_or(0.0);
    serialized.append(reinterpret_cast<const char*>(&creation_val), sizeof(double));
    size_t len = msg.origin.has_value() ? msg.origin->name.size() : 0;
    serialized.append(reinterpret_cast<const char*>(&len), sizeof(size_t));
    if (len > 0) serialized += msg.origin->name;
    len = msg.data.size();
    serialized.append(reinterpret_cast<const char*>(&len), sizeof(size_t));
    serialized += msg.data;
    if (serialized.size() > _channel.msg_size) {
        flock(_lock_fd, LOCK_UN);
        throw runtime_error("Message too large");
    }
    memcpy(_shm_ptr, serialized.data(), serialized.size());
    memset(static_cast<char*>(_shm_ptr) + serialized.size(), 0, _channel.msg_size - serialized.size());
    flock(_lock_fd, LOCK_UN);
}

void Publisher::close() {
    flock(_lock_fd, LOCK_EX);
    if (_shm_ptr) munmap(_shm_ptr, _channel.msg_size);
    close_shm(_shm_fd, _channel.name());
    if (_lock_fd != -1) ::close(_lock_fd);
    flock(_lock_fd, LOCK_UN);
}

// Subscriber implementation
Subscriber::Subscriber(const NodeID& node_id, const ChannelSpec& channel, function<void(const BaseMessage&)> callback)
    : _node_id(node_id), _channel(channel), _callback(std::move(callback)), _running(true) {
    _shm_fd = init_shm(_channel);
    _shm_ptr = mmap(nullptr, _channel.msg_size, PROT_READ | PROT_WRITE, MAP_SHARED, _shm_fd, 0);
    if (_shm_ptr == MAP_FAILED) throw runtime_error("mmap failed");
    _shm_id = get_shm_id(_shm_fd);
    _lock_fd = gen_lock(_channel.name());
    _last_msg = get_msg();
}

Subscriber::~Subscriber() {
    if (_shm_ptr) munmap(_shm_ptr, _channel.msg_size);
    if (_shm_fd != -1) ::close(_shm_fd);
}

optional<BaseMessage> Subscriber::wait_for_message() {
    while (_running) {
        flock(_lock_fd, LOCK_EX);
        int cur_shm_id = get_shm_id(_shm_fd);
        if (cur_shm_id == -1) {
            _running = false;
            flock(_lock_fd, LOCK_UN);
            return nullopt;
        }
        if (cur_shm_id != _shm_id) {
            flock(_lock_fd, LOCK_UN);
            throw runtime_error("SHM ID changed");
        }
        auto msg = get_msg();
        flock(_lock_fd, LOCK_UN);
        if (!msg || (_last_msg.has_value() && *msg == *_last_msg)) {
            this_thread::sleep_for(POLL_INTERVAL);
        } else {
            _last_msg = msg;
            return msg;
        }
    }
    return nullopt;
}

void Subscriber::listen() {
    if (!_callback) throw runtime_error("Callback required for listen");
    _listen_thread = thread([this]() {
        while (auto msg = wait_for_message()) {
            _callback(*msg);
            this_thread::sleep_for(chrono::duration<double>(0.0));
        }
    });
}

void Subscriber::close() {
    _running = false;
    if (_listen_thread.joinable()) _listen_thread.join();
    flock(_lock_fd, LOCK_EX);
    if (_shm_ptr) munmap(_shm_ptr, _channel.msg_size);
    close_shm(_shm_fd, _channel.name());
    if (_lock_fd != -1) ::close(_lock_fd);
    flock(_lock_fd, LOCK_UN);
}

optional<BaseMessage> Subscriber::get_msg() {
    try {
        char* buf = static_cast<char*>(_shm_ptr);
        BaseMessage msg;
        size_t offset = 0;
        double creation_val = 0.0;
        if (offset + sizeof(double) > _channel.msg_size) return nullopt;
        memcpy(&creation_val, buf + offset, sizeof(double));
        if (creation_val == 0.0) return nullopt; // treat empty
        msg.creation = creation_val;
        offset += sizeof(double);
        size_t len;
        if (offset + sizeof(size_t) > _channel.msg_size) return nullopt;
        memcpy(&len, buf + offset, sizeof(size_t));
        offset += sizeof(size_t);
        if (len > 0 && len < _channel.msg_size && offset + len <= _channel.msg_size) {
            msg.origin = NodeID{std::string(buf + offset, len)};
        }
        offset += len;
        if (offset + sizeof(size_t) > _channel.msg_size) return nullopt;
        memcpy(&len, buf + offset, sizeof(size_t));
        offset += sizeof(size_t);
        if (len > 0 && len < _channel.msg_size && offset + len <= _channel.msg_size) {
            msg.data.assign(buf + offset, len);
        } else {
            return nullopt;
        }
        return msg;
    } catch (...) {
        return nullopt;
    }
}

// Helper functions
static int get_shm_id(int shm_fd) {
    struct stat st;
    if (fstat(shm_fd, &st) == -1) return -1;
    return st.st_ino;
}

static void close_shm(int shm_fd, const string& shm_name) {
    ::close(shm_fd);
    std::string path = LOCK_DIR + std::string("ipc_shm_") + shm_name;
    ::unlink(path.c_str());
}

static int init_shm(ChannelSpec& channel) {
    std::string path = LOCK_DIR + std::string("ipc_shm_") + channel.name();
    int fd = ::open(path.c_str(), O_CREAT | O_RDWR, 0666);
    if (fd == -1) throw runtime_error("open shm file failed");
    if (ftruncate(fd, static_cast<off_t>(channel.msg_size)) == -1) {
        struct stat st;
        if (fstat(fd, &st) == 0 && st.st_size > 0) {
            channel.msg_size = static_cast<size_t>(st.st_size);
        } else {
            ::close(fd);
            ::unlink(path.c_str());
            throw runtime_error("ftruncate failed on shm file");
        }
    }
    return fd;
}

static int gen_lock(const string& shm_name) {
    string lock_path = LOCK_DIR + shm_name + LOCK_SUFFIX;
    int fd = open(lock_path.c_str(), O_CREAT | O_RDWR, 0666);
    if (fd == -1) throw runtime_error("open lock failed");
    return fd;
}
